### PR TITLE
[SPIR-V] Enabling SPIR-V builtin lookup in device SYCL mode

### PR DESCRIPTION
**Pull request by Naghasan**
This patch enables implicitly fdeclare-spirv-builtins when building sycl code in device mode and the runtime header changes related to flag being enabled. It also reverts some of the manual mangling done in libclc.

Signed-off-by: Victor Lomuller <victor@codeplay.com>

### CODE REVIEW

File: sycl/include/CL/__spirv/spirv_ops.hpp [hunk 1]
```diff
@@ -563,8 +251,6 @@ OpGroupAsyncCopyLocalToGlobal(__spv::Scope Execution, dataT *Dest, dataT *Src,
 
 extern void __spirv_ocl_prefetch(const char *Ptr, size_t NumBytes) noexcept;
 
-#endif // !__SYCL_DEVICE_ONLY__
-
 extern SYCL_EXTERNAL void __spirv_ControlBarrier(__spv::Scope Execution,
```
Comment (bader) [inline sycl/include/CL/__spirv/spirv_ops.hpp]: Will we emit the rest of SPIR-V built-ins definitions by clang?

NOTE: I have opened https://github.com/intel/llvm/pull/1373, which has conflicting changes.

**Comment by bader** [comment]
@Naghasan, could you resolve the conflicts, please?

File: sycl/include/CL/__spirv/spirv_ops.hpp (hunk 1 above)
Comment (Naghasan) [reply sycl/include/CL/__spirv/spirv_ops.hpp]: I'm not sure of what you mean, this is part is related to host

**Comment by Naghasan** [comment]
> @Naghasan, could you resolve the conflicts, please?

Done

**Comment by bader** [comment]
@Naghasan, any ideas why macro is not defined?
```
d:\buildbot\product_worker_intel\build_pr_with_lit_win\llvm.obj\bin\..\include\sycl\CL/__spirv/spirv_ops.hpp:25:5: error: invalid token at start of a preprocessor expression
#if not defined(__SPIRV_BUILTIN_DECLARATIONS__)
    ^
1 error generated.
error: command failed with exit status: 1
```

**Comment by Naghasan** [comment]
> any ideas why macro is not defined?

The macro is defined, the `#if not` is not portable. I changed it to a `#ifndef`

**Comment by bader** [comment]
> > any ideas why macro is not defined?
> 
> The macro is defined, the `#if not` is not portable. I changed it to a `#ifndef`

This seems to work, but there are still 10 failed LIT tests on Windows.

**Comment by Naghasan** [comment]
> > > any ideas why macro is not defined?
> > 
> > 
> > The macro is defined, the `#if not` is not portable. I changed it to a `#ifndef`
> 
> This seems to work, but there are still 10 failed LIT tests on Windows.

I neglected LLP64 vs LP64 ... I made a fix for this but limits inputs to be `int<len>_t` for now. As there is a systematic type check/conversion before calling the builtins that should be OK for now. But it is not perfect though, I'll work a follow up and more flexible solution.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -4086,6 +4086,7 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     // We want to compile sycl kernels.
     CmdArgs.push_back("-fsycl");
     CmdArgs.push_back("-fsycl-is-device");
+    CmdArgs.push_back("-fdeclare-spirv-builtins");
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp]: Driver changes look good to me.  Any reason why set this in the driver as opposed to in the FE?

**Comment by keryell** [comment]
> From http://ci.llvm.intel.com:8010/#/builders/28/builds/1203/steps/16/logs/FAIL__SYCL__aliases_cpp:
> 
> ```
> d:\buildbot\product_worker_intel\build_pr_with_lit_win\llvm.obj\bin\..\include\sycl\CL/__spirv/spirv_ops.hpp

This mix and match of `\` and `/` is a little bit scary at the first place...
But not related to this issue.

File: sycl/include/CL/__spirv/spirv_ops.hpp (hunk 1 above)
Comment (bader) [reply sycl/include/CL/__spirv/spirv_ops.hpp]: Sorry, I didn't notice that this is host-only code now.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (Naghasan) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Technically the change is not specific to SYCL so I designed it separately from SYCL and enable the feature via the driver.

**Pull request merged**